<h3>1. Atividade: Instruções</h3>
<p>Comece adicionando um título e escrevendo um parágrafo em uma célula de remarcação, apresentando o projeto e o conjunto de dados. O título e a introdução são provisórios neste momento; portanto, não gaste muito tempo aqui - você pode voltar no final do seu trabalho para refiná-los.</p>
<p>Leia o arquivo hacker_news.csv e coloque em uma lista de lista.<ul>
<li>Atribuao resultado a variável de nome <b>hn</b></li>
</ul>
<p>Mostre as primeiras cinco linhas do hn</p>

<p></p>

In [2]:
from csv import reader

In [3]:
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)
hn[0:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

<h3>2. Atividade: Instruções</h3>
<p>Extrai a primeira linha do data, e atribua a uma variavel chamada <b>headers</b>.</p>
<p>Remova a primeira linha do <b>hn</b>.</p>
<p>Mostre o <b>headers</b>.</p>
<p>Mostre as cinco primeiras linhas do hn para verificar se removeu da forma certa o header.</p>

In [4]:
headers = hn[0]
headers

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

In [5]:
hn= hn[1:]
hn[0:5]

[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12']]

<h3>3. Atividade: Instruções</h3>
<p>Crie 3 listas vazias e as chame de <b>ask_posts</b>, <b>show_posts</b>, e <b>other_posts</b>.</p>
<p>Percorra cada linha do <b>hn</b>.</p>
<ul>
<li>Atribua o título em cada linha a uma variável chamada <b>title</b>.</li>
<li><ul>
<li>Considerando que o title é segunda coluna,é necessário pegar o elemento do index 1 em cada linha.</li>
</ul></li>
</ul>
<p>Implementa os seguintes passos:</p>
<ul>
<li>Se a versão em minúscula do <b>title</b> começar com <b>ask hn</b>, anexe a linha a <b>ask_posts</b>. </li>
<li>Se não a versão minúscula do <b>title</b> começar com <b>show hn</b>, adicione a linha ao <b>show_posts</b>.</li>
<li>Se não adicione a linha ao <b>other_posts</b>.</li>
</ul>
<p>Verifique o número de postes em cada lista: <b>ask_posts</b>, <b>show_posts</b>, e <b>other_posts</b>.</p>

In [6]:
ask_posts = []
show_posts = []
other_posts = []
for row in hn:
    title = row[1].lower()   
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

In [7]:
print("Quantidade de ASK HN: ",len(ask_posts))
print("Quantidade de SHOW HN: ",len(show_posts))
print("Quantidade de OTHER POSTS: ",len(other_posts))
print("Quantidade do HN: ",len(hn))

Quantidade de ASK HN:  1744
Quantidade de SHOW HN:  1162
Quantidade de OTHER POSTS:  17194
Quantidade do HN:  20100


<h3>4. Atividade: Instruções</h3>
<p>Encontre o número total de comentários do ask posts e atribua-o a <b>total_ask_comments</b>.</p>
<ul>
<li>Defina <b>total_ask_comments</b> como 0.</li>
</ul>
<p>Use o for para percorrer ask posts.</p>
<dl>
<dt>Considerando que o <u>num_comments</u> é quinta coluna do <u>ask_posts</u>,é necessário pegar o elemento do index 4 em cada linha.</dt>
<dd>É necessário converter os valores em integer para poder calcular a soma de todos os comentários.</dd>
<dd>E adicione o valor ao <b>total_ask_comments</b>.</dd>
</dl>
<p>Calcule a média do número dos comentários do ask posts e atribua a variável<b>avg_ask_comments</b>.</p>
<p>Imprima a variável<b>avg_ask_comments</b>.</p>


<p>Encontre o número total de comentários do show posts e atribua-o a <b>total_show_comments</b>.</p>
<ul>
<li>Defina <b>total_show_comments</b> como 0.</li>
</ul>
<p>Use o for para percorrer show posts.</p>
<dl>
<dt>Considerando que o <u>num_comments</u> é quinta coluna do <u>show_posts</u>,é necessário pegar o elemento do index 4 em cada linha.</dt>
<dd>É necessário converter os valores em integer para poder calcular a soma de todos os comentários.</dd>
<dd>E adicione o valor ao <b>total_show_comments</b>.</dd>
</dl>
<p>Calcule a média do número dos comentários do show posts e atribua a variável<b>avg_show_comments</b>.</p>
<p>Imprima a variável<b>avg_show_comments</b>.</p>
<p>O show posts ou o ask posts recebe em média mais comentários?Escreva no markdown cell explicado suas descobertas.</p>

In [8]:
total_ask_comments = 0
for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments+=num_comments
avg_ask_comments = total_ask_comments/len(ask_posts)
print("A média dos comentários do ASK HN:",round(avg_ask_comments,2))

A média dos comentários do ASK HN: 14.04


In [9]:
total_show_comments = 0
for row in show_posts:
    num_comments = int(row[4])
    total_show_comments+=num_comments
avg_show_comments = total_show_comments/len(show_posts)
print("A média dos comentários do SHOW HN:",round(avg_show_comments,2))

A média dos comentários do SHOW HN: 10.32


<p> Em média o Ask posts tem mais comentários que o Show posts como mostrado acima.</p>

<h3>5. Atividade: Instruções</h3>
<p>Importe o módulo datetime como dt</p>
<p>Crie uma lista vázia e chama de <b>result_list</b>. Esta é uma lista de lista. </p>
<p>Percorra o <b>ask_posts</b> e adicione ao <b>result_list</b> a lista com dois elementos: </p>
<dl>
<dt>O primeiro elemento deverá ser a coluna <u>created_at.</u></dt>
<dd>Considerando que o <b>created_at</b> é a setima coluna do <b>ask_posts</b>,é necessário pegar o elemento do indes 6 de cada linha.</dd>
<dt>O segundo elemento deverá ser o número de comentários do post.</dt>
<dd>É necessário converter o valor em interger.</dd>
</dl>
<p>Criar dois dicionários vazios chamados <b>counts_by_hour</b> e <b>comments_by_hour</b>.</p>
<p>Percorra cada linha do <b>result_list</b>.</p>
<p>Extraia a hora da data,que é o primeiro elemento da linha. </p>
<p>Use o método <b>datetime.strptime()</b> e crie um objeto datetime.</p>
<p>Use o string que analisamos com primeiro argumento e string especifica o formato como o segundo argumento.</p>
<dl>
<dt>Use o método <u>datetime.strftime()</u>para selecionar apenas a hora do objeto datetime.</dt>
<dt>Se a hora não é um chave <u>counts_by_hour</u>:</dt>
<dd>Crie uma chave no <b>counts_by_hour</b> e coloque igual a 1.</dd>
<dd>Crie uma chave no <b>comments_by_hour</b> e coloque igual ao número do <b>comment</b>.</dd>
<dt>Se a hora já é uma chave no <u>counts_by_hour</u>:</dt>
<dd>Incremente o valor do <b>counts_by_hour</b> com 1.</dd>
<dd>Icremente o valor do <b>comments_by_hour</b> com o número do <b>comment</b>.</dd>
</dl>

In [10]:
import datetime as dt

In [11]:
result_list=[]
for row in ask_posts:
    created_at = row[6]
    num_comments = int(row[4])
    result_list.append([created_at,num_comments])
print(result_list[0])

['8/16/2016 9:55', 6]


In [12]:
counts_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M"
for row in result_list:
    date = row[0]
    date_str = dt.datetime.strptime(date,date_format)
    hour_str = date_str.strftime("%H")
    comments = row[1]
    if hour_str not in counts_by_hour:
        counts_by_hour[hour_str] = 1
        comments_by_hour[hour_str] = comments
    elif hour_str in counts_by_hour:
        counts_by_hour[hour_str] += 1
        comments_by_hour[hour_str] += comments


In [13]:
print(counts_by_hour)

{'12': 73, '04': 47, '18': 109, '22': 71, '10': 59, '06': 44, '15': 116, '02': 58, '23': 68, '19': 110, '21': 109, '16': 108, '05': 46, '01': 60, '13': 85, '20': 80, '07': 34, '17': 100, '03': 54, '00': 55, '11': 58, '08': 48, '09': 45, '14': 107}


In [14]:
print(comments_by_hour)

{'12': 687, '04': 337, '18': 1439, '22': 479, '10': 793, '06': 397, '15': 4477, '02': 1381, '23': 543, '19': 1188, '21': 1745, '16': 1814, '05': 464, '01': 683, '13': 1253, '20': 1722, '07': 267, '17': 1146, '03': 421, '00': 447, '11': 641, '08': 492, '09': 251, '14': 1416}


<h3>6. Atividade: Instruções</h3>
<p>Use o exemplo acima para calcular o número médio de comentários por postagem para postagens criadas durante cada hora do dia.</p>
<p>O resultado deve ser uma lista de listas nas quais o primeiro elemento é a hora e o segundo elemento é o número médio de comentários por postagem. Atribua o resultado a uma variável chamada <b>avg_by_hour</b>. Exiba os resultados.</p>


In [17]:
avg_by_hour = []

for comments in comments_by_hour:
    avg_by_hour.append([comments,(comments_by_hour[comments]/counts_by_hour[comments])])
print(avg_by_hour)

[['12', 9.41095890410959], ['04', 7.170212765957447], ['18', 13.20183486238532], ['22', 6.746478873239437], ['10', 13.440677966101696], ['06', 9.022727272727273], ['15', 38.5948275862069], ['02', 23.810344827586206], ['23', 7.985294117647059], ['19', 10.8], ['21', 16.009174311926607], ['16', 16.796296296296298], ['05', 10.08695652173913], ['01', 11.383333333333333], ['13', 14.741176470588234], ['20', 21.525], ['07', 7.852941176470588], ['17', 11.46], ['03', 7.796296296296297], ['00', 8.127272727272727], ['11', 11.051724137931034], ['08', 10.25], ['09', 5.5777777777777775], ['14', 13.233644859813085]]


<h3>7. Atividade: Instruções</h3>
<p>Crie uma lista que seja igual a <b>avg_by_hour</b> com colunas trocadas.</p>
<ul>
<li>Crie uma lista vazia e atribua-a a <b>swap_avg_by_hour.</b></li>
<li>Percorra as linhas de <b>avg_by_hour</b> e anexe a <b>swap_avg_by_hour.</b> cujo primeiro elemento seja o segundo elemento do <b>avg_by_hour</b> e segundo elemento seja o primeiro elemento do <b>avg_by_hour</b>.</li>
</ul>
<p>Imprima o <b>swap_avg_by_hour.</b>.</p>
<p>Use a função <b>sorted</b> para ordernar de forma descrecente o <b>swap_avg_by_hour</b>.Como a primeira coluna desta lista é o número médio de comentários, a classificação será ordenada pelo número médio de comentários.</p>
<ul>
<li>Defina o argumento <b>reverse</b> como <b>True</b>, para que o valor mais alto na seja o primeiro aparecer na lista.</li>
<li>Atribua o resultado a variavel<b>sorted_swap</b></li>
</ul>
<p>Imprima s string "Top 5 Hours for Ask Posts Comments".</p>
<p>Faça um loop em cada média e a cada hora (nesta ordem) nas cinco primeiras listas de <b>sorted_swap</b>.</p>
<p>Use o método srt.format() para imprimir as horas e as médias da seguinte forma: <b>15:00: 38.59 average comments per post.</b></p>
<ul>
<li>Para formatar as horas, use o construtor <b>datetime.strptime()</b> para retornar um objeto datetime e, em seguida, use o método <b>strftime()</b> para especificar o formato da hora.</li>
<li>Para formatar a média, você pode usar <b>{: .2f}</b> para indicar que apenas duas casas decimais devem ser usadas.</li>
</ul>
<p>Em que horas você deve criar uma postagem para ter uma chance maior de receber comentários? Consulte a documentação do conjunto de dados para converter os horários no fuso horário em que você vive. Escreva uma célula de marcação explicando suas descobertas.</p>

In [18]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])
print(swap_avg_by_hour) 

[[9.41095890410959, '12'], [7.170212765957447, '04'], [13.20183486238532, '18'], [6.746478873239437, '22'], [13.440677966101696, '10'], [9.022727272727273, '06'], [38.5948275862069, '15'], [23.810344827586206, '02'], [7.985294117647059, '23'], [10.8, '19'], [16.009174311926607, '21'], [16.796296296296298, '16'], [10.08695652173913, '05'], [11.383333333333333, '01'], [14.741176470588234, '13'], [21.525, '20'], [7.852941176470588, '07'], [11.46, '17'], [7.796296296296297, '03'], [8.127272727272727, '00'], [11.051724137931034, '11'], [10.25, '08'], [5.5777777777777775, '09'], [13.233644859813085, '14']]


In [21]:
sorted_swap = sorted(swap_avg_by_hour,reverse=True)
print(sorted_swap)

[[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [13.20183486238532, '18'], [11.46, '17'], [11.383333333333333, '01'], [11.051724137931034, '11'], [10.8, '19'], [10.25, '08'], [10.08695652173913, '05'], [9.41095890410959, '12'], [9.022727272727273, '06'], [8.127272727272727, '00'], [7.985294117647059, '23'], [7.852941176470588, '07'], [7.796296296296297, '03'], [7.170212765957447, '04'], [6.746478873239437, '22'], [5.5777777777777775, '09']]


In [23]:
print( "Top 5 Hours for Ask Posts Comments")
hour_format = "%H"
for row in sorted_swap:
    hour = row[1]
    por = row[0]
    hour_dt = dt.datetime.strptime(hour,hour_format)
    hour_str = hour_dt.strftime("%H:%M")
    print("{}: {:.2f} average comments per post".format(hour_str,por))
    

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post
13:00: 14.74 average comments per post
10:00: 13.44 average comments per post
14:00: 13.23 average comments per post
18:00: 13.20 average comments per post
17:00: 11.46 average comments per post
01:00: 11.38 average comments per post
11:00: 11.05 average comments per post
19:00: 10.80 average comments per post
08:00: 10.25 average comments per post
05:00: 10.09 average comments per post
12:00: 9.41 average comments per post
06:00: 9.02 average comments per post
00:00: 8.13 average comments per post
23:00: 7.99 average comments per post
07:00: 7.85 average comments per post
03:00: 7.80 average comments per post
04:00: 7.17 average comments per post
22:00: 6.75 average comments per post
09:00: 5.58 average comments per post


<p>Os melhores horários para se criar um post no Hacker News 15:00, 02:00, 20:00, 16:00, 21:00 são horários do .</p>
<table border="1">
<tr>
<th>Fuso Horário do Leste dos EUA</th>
<th>Fuso Horário de São Paulos</th>
</tr>
<tr>
<td>15:00</td>
<td>16:00</td>
</tr>
<tr>
<td>02:00</td>
<td>03:00</td>
</tr>
<tr>
<td>20:00</td>
<td>21:00</td>
</tr>
<tr>
<td>16:00</td>
<td>17:00</td>
</tr>
<tr>
<td>21:00 </td>
<td>22:00</td>
</tr>
</table>